### PR TITLE
Upgraded: Python3 compatibility (contd)

### PR DESCRIPTION
**Pull request by agamdua**
- Added `six` as a dependency to handle `iter*()` for dictionary operations
 - `filter` is lazy in py3, added a list to expand it. should work as is in py2
   and avoid breaking API changes

### CODE REVIEW

**Comment by thieman** [comment]
I'm fine with adding Py3 support but I hate adding the `six` library to get such a tiny, simple function. Please either copy the relevant `iter` methods out of `six` and provide attribution (I think you need to copy their license) or provide your own implementation of that functionality.

**Comment by agamdua** [comment]
Has been updated with commits from downstream that remove the `six` dependency.

File: setup.py [hunk 1]
```diff
@@ -1,12 +1,16 @@
...
+
+install_requires = []
+
 if python_ver.major == 2 and python_ver.minor < 7:
-    install_requires = ['ordereddict']
+    install_requires += ['ordereddict']
 
 setup(name='py-dag',
-      version='2.5.0',
+      version='2.6.0',
```
Comment (thieman) [inline setup.py]: This should not be part of this PR

File: tests/init_tests.py [hunk 2]
```diff
@@ -43,44 +48,52 @@ def test_from_dict():
...
+
 @with_setup(start_with_graph)
 def test_downstream():
     assert set(dag.downstream('a', dag.graph)) == set(['b', 'c'])
 
+
 @with_setup(start_with_graph)
 def test_all_downstreams():
-    assert dag.all_downstreams('a') == ['c', 'b', 'd']
+    assert sorted(dag.all_downstreams('a')) == ['b', 'c', 'd']
```
Comment (thieman) [inline tests/init_tests.py]: You've made this test worthless, the intention of `all_downstreams` is that it returns in topological order (refer to its docstring). The assert needs to be restored to its original form, and your PR needs to pass that test.

File: tests/init_tests.py [hunk 3]
```diff
@@ -43,44 +48,52 @@ def test_from_dict():
...
     dag.from_dict({'a': [],
                    'b': ['a'],
                    'c': ['b']})
     assert dag.topological_sort() == ['c', 'b', 'a']
 
+
 @with_setup(start_with_graph)
 def test_successful_validation():
-    assert dag.validate()[0] == True
+    assert dag.validate()[0] is True
```
Comment (thieman) [inline tests/init_tests.py]: Good catch

File: dag/six_subset.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import operator
+import sys
+
+# Useful for very coarse version differentiation.
+PY2 = sys.version_info[0] == 2
+PY3 = sys.version_info[0] == 3
```
Comment (thieman) [inline dag/six_subset.py]: I believe this is more future-proof as `>= 3` given GVR's dumb comments about 4.0 being the version that follows 3.9.

File: tests/init_tests.py (hunk 2 above)
Comment (agamdua) [reply tests/init_tests.py]: cheating cheating cheating

File: dag/six_subset.py (hunk 4 above)
Comment (agamdua) [reply dag/six_subset.py]: agreed

File: tests/init_tests.py (hunk 2 above)
Comment (agamdua) [reply tests/init_tests.py]: also this is odd, all I changed was add a `list` over the `filter` call. That should not affect the ordering, let me check if this is a side effect from another change.

File: dag/__init__.py [hunk 5]
```diff
@@ -1,9 +1,11 @@
 from copy import copy, deepcopy
 from collections import deque
 
+from . import six_subset as six
+
 try:
     from collections import OrderedDict
-except:
+except ImportError:
```
Comment (thieman) [inline dag/__init__.py]: Good change, thank you

File: .travis.yml [hunk 6]
```diff
@@ -3,5 +3,8 @@ language: python
 python:
   - "2.7"
   - "2.6"
+  - "3.4"
```
Comment (thieman) [inline .travis.yml]: Any actual reason for not including <=3.3?

Comment (agamdua) [reply .travis.yml]: No one uses < 3.3, there is no specific reason for not including 3.3, I can add it.

Comment (thieman) [reply .travis.yml]: Eh if you think penetration is actually low I'm fine not wasting the CI cycles on it

File: tests/init_tests.py [hunk 7]
```diff
@@ -92,17 +105,20 @@ def test_all_downstreams_pass_graph():
     assert dag.all_downstreams('b', dag2.graph) == ['d']
     assert dag.all_downstreams('d', dag2.graph) == []
 
+
 @with_setup(start_with_graph)
 def test_predecessors():
     assert set(dag.predecessors('a')) == set([])
-    assert set(dag.predecessors('b')) == set(['a'])
```
Comment (agamdua) [inline tests/init_tests.py]: these updates were necessary since I modified the graph, refer the message https://github.com/thieman/py-dag/pull/23/commits/d5026f9f6347417077bf1add468006a2142a7389

**Comment by thieman** [comment]
Mergin', thanks for the patch. Will cut a new version shortly.
